third/nmh/man/pick.man
.\"
.\" %nmhwarning%
.\" $Id: pick.man,v 1.1.1.1 1999-02-07 18:14:21 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH PICK %manext1% MH.6.8 [%nmhversion%]
.SH NAME
pick \- search for messages by content
.SH SYNOPSIS
.in +.5i
.ti -.5i
pick
\%[+folder] \%[msgs]
\%[\-and\ ...] \%[\-or\ ...] \%[\-not\ ...]
.br
\%[\-lbrace\ ...\ \-rbrace]
\%[\-\|\-component\ pattern]
.br
\%[\-cc\ pattern]
\%[\-date\ pattern]
\%[\-from\ pattern]
.br
\%[\-search\ pattern]
\%[\-subject\ pattern]
\%[\-to\ pattern]
.br
\%[\-after\ date]
\%[\-before\ date]
\%[\-datefield\ field]
.br
\%[\-sequence\ name\ ...]
\%[\-public] \%[\-nopublic]
\%[\-zero]
.br
\%[\-nozero]
\%[\-list] \%[\-nolist]
\%[\-version]
\%[\-help]
.ti .5i

typical usage:
.br
scan\0`pick\0\-from\0jones`
.br
pick\0\-to\0holloway\0\-sequence\0select
.br
show\0`pick\0\-before\0friday`
.in -.5i
.SH DESCRIPTION
\fIPick\fR searches within a folder for messages with the specified
contents, and then identifies those messages.  Two types of search
primitives are available: pattern matching and date constraint
operations.

A modified \fIgrep\fR(1) is used to perform the matching, so the
full regular expression (see \fIed\fR(1)) facility is available
within `pattern'.  With `\-search', `pattern' is used directly,
and with the others, the grep pattern constructed is:

.ti +.5i
\*(lqcomponent[ \\t]*:\&.*pattern\*(rq

This means that the pattern specified for a `\-search' will be found
everywhere in the message, including the header and the body, while
the other pattern matching requests are limited to the single specified
component.  The expression

.ti +.5i
`\-\|\-component\ pattern'

is a shorthand for specifying

.ti +.5i
`\-search \*(lqcomponent[ \\t]*:\&.*pattern\*(rq\ '

It is used to pick a component which is not one of \*(lqTo:\*(rq,
\*(lqcc:\*(rq, \*(lqDate:\*(rq, \*(lqFrom:\*(rq, or \*(lqSubject:\*(rq.
An example is `pick\0\-\|\-reply\-to\0pooh'.

Pattern matching is performed on a per\-line basis.  Within the header
of the message, each component is treated as one long line, but in the
body, each line is separate.  Lower\-case letters in the search pattern
will match either lower or upper case in the message, while upper case
will match only upper case.

Note that since the `\-date' switch is a pattern matching operation (as
described above), to find messages sent on a certain date the pattern
string must match the text of the \*(lqDate:\*(rq field of the message.

Independent of any pattern matching operations requested, the switches
`\-after date' or `\-before date' may also be used to introduce date/time
constraints on all of the messages.  By default, the \*(lqDate:\*(rq
field is consulted, but if another date yielding field (such as
\*(lqBB\-Posted:\*(rq or \*(lqDelivery\-Date:\*(rq) should be used, the
`\-datefield\ field' switch may be used.

With `\-before' and `\-after', \fIpick\fR will actually parse the date
fields in each of the messages specified in `msgs' and compare them
to the date/time specified.  If `\-after' is given, then only those
messages whose \*(lqDate:\*(rq field value is chronologically after the
date specified will be considered.  The `\-before' switch specifies the
complimentary action.

Both the `\-after' and `\-before' switches take legal 822\-style date
specifications as arguments.  \fIPick\fR will default certain missing
fields so that the entire date need not be specified.  These fields
are (in order of defaulting): timezone, time and timezone, date, date
and timezone.  All defaults are taken from the current date, time,
and timezone.

In addition to 822\-style dates, \fIpick\fR will also recognize any of
the days of the week (\*(lqsunday\*(rq, \*(lqmonday\*(rq, and so on),
and the special dates \*(lqtoday\*(rq, \*(lqyesterday\*(rq (24 hours
ago), and \*(lqtomorrow\*(rq (24 hours from now).  All days of the
week are judged to refer to a day in the past (e.g., telling \fIpick\fR
\*(lqsaturday\*(rq on a \*(lqtuesday\*(rq means \*(lqlast\ saturday\*(rq
not \*(lqthis\ saturday\*(rq).

Finally, in addition to these special specifications, \fIpick\fR will
also honor a specification of the form \*(lq\-dd\*(rq, which means
\*(lqdd days ago\*(rq.

\fIPick\fR supports complex boolean operations on the searching primitives
with the `\-and', `\-or', `\-not', and `\-lbrace\ ...\ \-rbrace' switches.
For example,

.ti +.5i
.ie t \{\
pick\0\-after\0yesterday\0\-and\0\-lbrace\0\-from\0freida\0\-or\0\-from\0fear\0\-rbrace
.\}
.el \{\
pick\0\-after\0yesterday\0\-and
.br
.ti +1i
\-lbrace\0\-from\0freida\0\-or\0\-from\0fear\0\-rbrace
.\}

identifies messages recently sent by \*(lqfrieda\*(rq or \*(lqfear\*(rq.

The matching primitives take precedence over the `\-not' switch, which
in turn takes precedence over `\-and' which in turn takes precedence
over `\-or'.  To override the default precedence, the `\-lbrace' and
`\-rbrace' switches are provided, which act just like opening and closing
parentheses in logical expressions.

If no search criteria are given, all the messages specified on the
command line are selected (this defaults to \*(lqall\*(rq).

Once the search has been performed, if the `\-list' switch is given, the
message numbers of the selected messages are written to the standard
output separated by newlines.  This is \fIextremely\fR useful for
quickly generating arguments for other \fInmh\fR programs by using the
\*(lqbackquoting\*(rq syntax of the shell.  For example, the command

.ti +.5i
scan\0`pick\0+todo\0\-after\0\*(lq31 Mar 83 0123 PST\*(rq`

says to \fIscan\fR those messages in the indicated folder which meet the
appropriate criterion.  Note that since \fIpick\fR\0's context changes
are written out prior to \fIscan\fR\0's invocation, you need not give
the folder argument to \fIscan\fR as well.

Regardless of the operation of the `\-list' switch, the `\-sequence name'
switch may be given once for each sequence the user wishes to define.
For each sequence named, that sequence will be defined to mean exactly
those messages selected by \fIpick\fR.  For example,

.ti +.5i
pick\0\-from\0frated\0\-seq\0fred

defines a new message sequence for the current folder called
\*(lqfred\*(rq which contains exactly those messages that were selected.

Note that whenever \fIpick\fR processes a `\-sequence\ name' switch, it
sets `\-nolist'.

By default, \fIpick\fR will zero the sequence before adding it.  This
action can be disabled with the `\-nozero' switch, which means that the
messages selected by \fIpick\fR will be added to the sequence, if it
already exists, and any messages already a part of that sequence will
remain so.

The `\-public' and `\-nopublic' switches are used by \fIpick\fR in the
same way \fImark\fR uses them.
.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Current\-Folder:~^To find the default current folder
.Sa
mark(1)
.De
`+folder' defaults to the current folder
.Ds
`msgs' defaults to all
.Ds
`\-datefield date'
.Ds
`\-zero'
.Ds
`\-list' is the default if no `\-sequence', `\-nolist' otherwise
.Co
If a folder is given, it will become the current folder.
.Hi
In previous versions of \fIMH\fR, the \fIpick\fR command would \fIshow\fR,
\fIscan\fR, or \fIrefile\fR the selected messages.  This was rather
\*(lqinverted logic\*(rq from the UNIX point of view, so \fIpick\fR was
changed to define sequences and output those sequences.  Hence, \fIpick\fR
can be used to generate the arguments for all other \fIMH\fR commands,
instead of giving \fIpick\fR endless switches for invoking those commands
itself.

Also, previous versions of \fIpick\fR balked if you didn't specify
a search string or a date/time constraint.  The current version does
not, and merely matches the messages you specify.  This lets you type
something like:

.ti +.5i
show\0`pick\0last:20\0\-seq\0fear`

instead of typing

.in +.5i
.nf
mark\0\-add\0\-nozero\0\-seq\0fear\0last:20
show\0fear
.fi
.in -.5i

Finally, timezones used to be ignored when comparing dates: they aren't
any more.
.Hh
Use \*(lqpick sequence \-list\*(rq to enumerate the messages in a sequence
(such as for use by a shell script).
.Bu
The argument to the `\-after' and `\-before' switches must be interpreted
as a single token by the shell that invokes \fIpick\fR.  Therefore, one
must usually place the argument to this switch inside double\-quotes.
Furthermore, any occurrence of `\-datefield' must occur prior to the
`\-after' or `\-before' switch it applies to.

If \fIpick\fR is used in a back\-quoted operation, such as

.ti +.5i
scan\0`pick\0\-from\0jones`

and \fIpick\fR selects no messages (e.g., no messages are from
\*(lqjones\*(rq), then the shell will still run the outer command (e.g.,
\*(lqscan\*(rq).  Since no messages were matched, \fIpick\fR produced
no output, and the argument given to the outer command as a result of
backquoting \fIpick\fR is empty.  In the case of \fInmh\fR programs,
the outer command now acts as if the default `msg' or `msgs' should be
used (e.g., \*(lqall\*(rq in the case of \fIscan\fR\0).  To prevent this
unexpected behavior, if `\-list' was given, and if its standard output is
not a tty, then \fIpick\fR outputs the illegal message number \*(lq0\*(rq
when it fails.  This lets the outer command fail gracefully as well.

.sp
The pattern syntax \*(lq[l-r]\*(rq is not supported; each letter to be
matched must be included within the square brackets.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/pick.c

/*
 * pick.c -- search for messages by content
 *
 * $Id: pick.c,v 1.1.1.1 1999-02-07 18:14:15 danw Exp $
 */

#include <h/mh.h>
#include <zotnet/tws/tws.h>
#include <h/picksbr.h>

/*
 * We allocate space for messages (msgs array)
 * this number of elements at a time.
 */
#define MAXMSGS  256


static struct swit switches[] = {
#define	ANDSW                   0
    { "and", 0 },
#define	ORSW                    1
    { "or", 0 },
#define	NOTSW                   2
    { "not", 0 },
#define	LBRSW                   3
    { "lbrace", 0 },
#define	RBRSW                   4
    { "rbrace", 0 },
#define	CCSW                    5
    { "cc  pattern", 0 },
#define	DATESW                  6
    { "date  pattern", 0 },
#define	FROMSW                  7
    { "from  pattern", 0 },
#define	SRCHSW                  8
    { "search  pattern", 0 },
#define	SUBJSW                  9
    { "subject  pattern", 0 },
#define	TOSW                   10
    { "to  pattern", 0 },
#define	OTHRSW                 11
    { "-othercomponent  pattern", 0 },
#define	AFTRSW                 12
    { "after date", 0 },
#define	BEFRSW                 13
    { "before date", 0 },
#define	DATFDSW                14
    { "datefield field", 5 },
#define	SEQSW                  15
    { "sequence name", 0 },
#define	PUBLSW                 16
    { "public", 0 },
#define	NPUBLSW                17
    { "nopublic", 0 },
#define	ZEROSW                 18
    { "zero", 0 },
#define	NZEROSW                19
    { "nozero", 0 },
#define	LISTSW                 20
    { "list", 0 },
#define	NLISTSW                21
    { "nolist", 0 },
#define VERSIONSW              22
    { "version", 0 },
#define	HELPSW                 23
    { "help", 4 },
    { NULL, 0 }
};

static int listsw = 0;


int
main (int argc, char **argv)
{
    int publicsw = -1, zerosw = 1, seqp = 0, vecp = 0;
    int nummsgs, maxmsgs, lo, hi, msgnum;
    char *maildir, *folder = NULL, buf[100];
    char *cp, **argp, **arguments;
    char **msgs, *seqs[NUMATTRS + 1], *vec[MAXARGS];
    struct msgs *mp;
    register FILE *fp;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /*
     * Allocate the initial space to record message
     * names, ranges, and sequences.
     */
    nummsgs = 0;
    maxmsgs = MAXMSGS;
    if (!(msgs = (char **) malloc ((size_t) (maxmsgs * sizeof(*msgs)))))
	adios (NULL, "unable to allocate storage");

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    if (*++cp == '-') {
		vec[vecp++] = --cp;
		goto pattern;
	    }
	    switch (smatch (cp, switches)) {
	    case AMBIGSW: 
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW: 
		adios (NULL, "-%s unknown", cp);

	    case HELPSW: 
		snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			  invo_name);
		print_help (buf, switches, 1);
		listsw = 0;	/* HACK */
		done (1);
	    case VERSIONSW:
		print_version(invo_name);
		done (1);

	    case CCSW: 
	    case DATESW: 
	    case FROMSW: 
	    case SUBJSW: 
	    case TOSW: 
	    case DATFDSW: 
	    case AFTRSW: 
	    case BEFRSW: 
	    case SRCHSW: 
		vec[vecp++] = --cp;
	    pattern:
		if (!(cp = *argp++))/* allow -xyz arguments */
		    adios (NULL, "missing argument to %s", argp[-2]);
		vec[vecp++] = cp;
		continue;
	    case OTHRSW: 
		adios (NULL, "internal error!");

	    case ANDSW:
	    case ORSW:
	    case NOTSW:
	    case LBRSW:
	    case RBRSW:
		vec[vecp++] = --cp;
		continue;

	    case SEQSW: 
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);

		/* check if too many sequences specified */
		if (seqp >= NUMATTRS)
		    adios (NULL, "too many sequences (more than %d) specified", NUMATTRS);
		seqs[seqp++] = cp;
		listsw = 0;
		continue;
	    case PUBLSW: 
		publicsw = 1;
		continue;
	    case NPUBLSW: 
		publicsw = 0;
		continue;
	    case ZEROSW: 
		zerosw++;
		continue;
	    case NZEROSW: 
		zerosw = 0;
		continue;

	    case LISTSW: 
		listsw++;
		continue;
	    case NLISTSW: 
		listsw = 0;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	} else {
	    /*
	     * Check if we need to allocate more space
	     * for message name/ranges/sequences.
	     */
	    if (nummsgs >= maxmsgs) {
		maxmsgs += MAXMSGS;
		if (!(msgs = (char **) realloc (msgs,
						(size_t) (maxmsgs * sizeof(*msgs)))))
		    adios (NULL, "unable to reallocate msgs storage");
	    }
	    msgs[nummsgs++] = cp;
	}
    }
    vec[vecp] = NULL;

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /*
     * If we didn't specify which messages to search,
     * then search the whole folder.
     */
    if (!nummsgs)
	msgs[nummsgs++] = "all";

    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder)))
	adios (NULL, "unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	adios (NULL, "no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < nummsgs; msgnum++)
	if (!m_convert (mp, msgs[msgnum]))
	    done (1);
    seq_setprev (mp);	/* set the previous-sequence */

    /*
     * If we aren't saving the results to a sequence,
     * we need to list the results.
     */
    if (seqp == 0)
	listsw++;

    if (publicsw == 1 && is_readonly(mp))
	adios (NULL, "folder %s is read-only, so -public not allowed", folder);

    if (!pcompile (vec, NULL))
	done (1);

    lo = mp->lowsel;
    hi = mp->hghsel;

    /*
     * Scan through all the SELECTED messages and check for a
     * match.  If the message does not match, then unselect it.
     */
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	if (is_selected (mp, msgnum)) {
	    if ((fp = fopen (cp = m_name (msgnum), "r")) == NULL)
		admonish (cp, "unable to read message");
	    if (fp && pmatches (fp, msgnum, 0L, 0L)) {
		if (msgnum < lo)
		    lo = msgnum;
		if (msgnum > hi)
		    hi = msgnum;
	    } else {
		/* if it doesn't match, then unselect it */
		unset_selected (mp, msgnum);
		mp->numsel--;
	    }
	    if (fp)
		fclose (fp);
	}
    }

    mp->lowsel = lo;
    mp->hghsel = hi;

    if (mp->numsel <= 0)
	adios (NULL, "no messages match specification");

    seqs[seqp] = NULL;

    /*
     * Add the matching messages to sequences
     */
    for (seqp = 0; seqs[seqp]; seqp++)
	if (!seq_addsel (mp, seqs[seqp], publicsw, zerosw))
	    done (1);

    /*
     * Print the name of all the matches
     */
    if (listsw) {
	for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
	    if (is_selected (mp, msgnum))
		printf ("%s\n", m_name (msgnum));
    } else {
	printf ("%d hit%s\n", mp->numsel, mp->numsel == 1 ? "" : "s");
    }

    context_replace (pfolder, folder);	/* update current folder         */
    seq_save (mp);			/* synchronize message sequences */
    context_save ();			/* save the context file         */
    folder_free (mp);			/* free folder/message structure */
    done (0);
}


void
done (int status)
{
    if (listsw && status && !isatty (fileno (stdout)))
	printf ("0\n");
    exit (status);
}
